File: LeetCode/Interview/160寻找两个链表的交点/intersection.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def intersection(self, headA: ListNode, headB: ListNode) -> ListNode:
        if not headA or not headB:
            return None

        a, b = headA, headB

        while a != b:
            a = headB if a is None else a.next
            b = headA if b is None else b.next

        return a


if __name__ == '__main__':
    a1, a2, a3 = ListNode(1), ListNode(2), ListNode(3)
    b1 = ListNode(4)
    c1, c2, c3 = ListNode(5), ListNode(6), ListNode(7)

    a1.next, a2.next, a3.next = a2, a3, c1
    b1.next = c1
    c1.next, c2.next = c2, c3

    print(Solution().intersection(a1, b1).val)

File: LeetCode/Interview/160寻找两个链表的交点/intersection.c
/******************************************************************************
 *
 * Name: intersection.c - Description
 * Created on 2025/02/28
 * Copyright (C) 2022 - 2025, wyh.
 *
 *****************************************************************************/

#include <stdio.h>

struct ListNode {
	int val;
	struct ListNode *next;
};

struct ListNode *getIntersectionNode(struct ListNode *headA,
				     struct ListNode *headB)
{
	if (!headA || !headB)
		return NULL;

	struct ListNode *a = headA;
	struct ListNode *b = headB;

	// 当 a = b 时，说明找到交点，如果二者最后都为 NULL，说明没有交点
	while (a != b) {
		a = ((a == NULL) ? headB : a->next);
		b = ((b == NULL) ? headA : b->next);
	}

	return a;
}
